Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp

```diff
@@ -73,7 +80,12 @@ WebNotificationManager::~WebNotificationManager()
 void WebNotificationManager::initialize(const WebProcessCreationParameters& parameters)
 {
 #if ENABLE(NOTIFICATIONS)
-    m_permissionsMap = parameters.notificationPermissions;
+    auto& webPermissionController = downcast<WebPermissionController>(PermissionController::shared());
+    for (auto const& [originString, allowed] : parameters.notificationPermissions) {
+        auto topOrigin = SecurityOrigin::createFromString(originString)->data();
+        auto permissionState = allowed ? PermissionState::Granted : PermissionState::Denied;
+        webPermissionController.updateCache(ClientOrigin { topOrigin, SecurityOriginData { } }, PermissionDescriptor { PermissionName::Notifications }, permissionState);
+    }
```

Comment:
Why this change? I thought we only want one implementation for both APIs, i.e. we will only use queryPermission to pull notification's permission.
This makes the cache dependent on WKNotificationManager: e.g. if client implements queryPermission but not WKNotificationManager, the result will be wrong (because cache will not be updated).

Reply:
This patch contains other code changes that ensure that calling `queryPermission` will also update the cache (if the permission state being queried is the state of the Notifications API.

The reason for this change was that in a previous version of this patch, the EWS bots detected a API test failure on GTK, which was a regression. The failure was that `Notification::permission` was incorrectly returning "default" instead of "granted". It turns out that the test client did not implement queryPermission and the test had been succeeding because of the cache in `WebNotificationManager`. Since this patch removes that cache and refactors `Notification::permission` to use `queryPermission`, the result was coming back as default. There were two options to fix this:

1. Implement `queryPermission` in the GTK test client
2. Make the not-working WebProcess cache in `WebPermissionController` work for notifications. 

The second option seemed easier and more beneficial since we'll have to do this for the on change event later on anyways. I wasn't sure how to test if this will fix the API test since it was a GTK failure, so I guess we'll have to wait for the bots to check. :(

Reply:
> This patch contains other code changes that ensure that calling queryPermission will also update the cache (if the permission state being queried is the state of the Notifications API.

Let's say client implements queryPermission but not WKNotificationManager, and user sets allowed notification (e.g. in per-site settings in Safari) for a site.
For the first permission::query, client returns allowed, and WebPermissionController caches the result. Then user decides to change the setting to denied (no longer wants to receive notification from the site), and tells WebKit client. For the second permission query, what would be returned with your patch?

IIUC, before this patch, notification.permission relies on WKNotificationManager and permission.query relies on queryPermission. After this patch, both rely on either WKNotificationManager or queryPermission + WKNotificationManager, which does not seem like a progress to me.

> The second option seemed easier and more beneficial since we'll have to do this for the on change event later on anyways.

The reason we implement this before change event is because we think this change would work on its own. If this change relies on cache to work, then I think we should implement change event first,  since the cache might not be in this form (an origin-to-permission map). You can imagine having WebPermissionController track PermissionStatus (or PermissionObserver) objects which hold the state, then you will need to revamp the cache.